swarm/network/discovery_test.go

```diff
@@ -58,3 +67,154 @@ func TestDiscovery(t *testing.T) {
+	registerBzzAddr(3, hive, false) // add a known but not connected peer
+	registerBzzAddr(1, hive, false) // add a known but not connected peer
+
+	// start the hive
+	hive.Start(s.Server)
+	defer hive.Stop()
+
+	// the remote node is the only one from the ProtocolTester
+	remote := s.Nodes[0]
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	// we need to wait until the remote node is actually connected to our hive
+WAIT_PIVOT:
+	for {
+		select {
+		case <-ctx.Done():
+			t.Fatal("Timed out waiting for the remote node to connect")
+		case <-time.After(100 * time.Millisecond):
+			if _, ok := hive.peers[remote.ID()]; ok {
+				break WAIT_PIVOT
+			}
+		}
+	}
+
+	// get BzzAddr of the remote
+	remoteAddress := hive.peers[remote.ID()]
+	remoteBzz := remoteAddress.BzzAddr.Over()
+
+	// now we need to identify which peers are expected
+	// iterate the hive's connection and only add peers below testDepth
+	var expectedPeers []*BzzAddr
+	hive.EachConn(remoteBzz, 255, func(p *Peer, po int) bool {
+		if po < testDepth {
+			return false
+		}
+		// don't add the remote node itself to expectedPeers;
+		// the remote node was not added as
+		if !bytes.Equal(p.BzzAddr.Over(), remoteBzz) {
+			expectedPeers = append(expectedPeers, p.BzzAddr)
+		}
+		return true
+	})
+
+	// the test exchange is as follows:
+	// 1. Trigger a subPeersMsg from remote to our hive
+	// 2. Hive will respond with peersMsg with the set of expected peers
+	err = s.TestExchanges(p2ptest.Exchange{
+		Label: "incoming subPeersMsg",
+		Expects: []p2ptest.Expect{
+			{
+				Code: 1,
+				Msg:  &subPeersMsg{Depth: uint8(2)},
+				Peer: remote.ID(),
+			},
+		},
+	},
+		p2ptest.Exchange{
+			Label: "trigger subPeers and receive peersMsg",
+			Triggers: []p2ptest.Trigger{
+				{
+					Code:    1,
+					Msg:     &subPeersMsg{Depth: uint8(testDepth)},
+					Peer:    remote.ID(),
+					Timeout: 3 * time.Second,
+				},
+			},
+			Expects: []p2ptest.Expect{
+				{
+					Code:    0,
+					Msg:     &peersMsg{Peers: expectedPeers},
+					Peer:    remote.ID(),
+					Timeout: 3 * time.Second,
+				},
+			},
+		})
+
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// add the BzzAddr to the hive
+func registerBzzAddr(po int, hive *Hive, on bool) {
+	a := pot.RandomAddressAt(pot.NewAddressFromBytes(hive.BaseAddr()), po)
+	bzzAddr := &BzzAddr{OAddr: a.Bytes(), UAddr: a.Bytes()}
+	if on {
+		// actually connect
+		peer := newDiscPeer(bzzAddr, a.String(), hive)
+		hive.On(peer)
+	} else {
+		// only add to address book
+		hive.Register(bzzAddr)
+	}
+}
+
+// as we are not creating a real node via the protocol,
+// we need to create the discovery peer objects for the additional kademlia
+// nodes manually
+func newDiscPeer(bzzAddr *BzzAddr, name string, hive *Hive) *Peer {
+	p2pPeer := p2p.NewPeer(adapters.RandomNodeConfig().ID, name, nil)
+	return NewPeer(&BzzPeer{
+		Peer:      protocols.NewPeer(p2pPeer, &dummyMsgRW{}, DiscoverySpec),
```

Comment:
just add nil for 2nd arg

Reply:
no, that doesn't work, on `hive.Stop()` this will produce a null pointer dereference